### ISSUE TITLE
Feature Request: Assign metadata to arms.

### ISSUE THREAD

**Issue by mrthekid** [facebook/Ax#640]
I am trying to do some online optimization following the modular/devloper-api tutorials online.  It seems, however, that most of these tutorials assume that 1) the trial you generate is the one that you evaluate and 2) that you can compute the objective from the arms somehow.  Taking the booth metric example (https://ax.dev/tutorials/building_blocks.html) for instance, this assumes that I have some programmatic way of getting the result from the arm parameters.  

In actual practice, I have no way of guaranteeing that the trial I generate is actually pushed to prod without modification, so I'm always just observing results.  (Much like the "attach_trial / complete_trial" paradigm in the service API.) Moreover, I can't compute the objective from the arms except perhaps by just storing the (arm, result) pairs somewhere else and then looking these up.  However, that seems very inelegant.  What would be nice is if I could just attach the result directly to the arms.  I see trials allow for metadata - why not arms? 

Taking the BoothMetric as an example - what I'd like to do in the for loop is just grab the objective mean + sem from the arm (e.g., mean  = arm.metadata[objective_name]) having previously recorded it.  I can do this with the trial itself, though, not at the arm level.  (I could of course attach some metadata to the trial that has information about all the arms, but this seems inelegant.)  

```
class BoothMetric(Metric):
    def fetch_trial_data(self, trial):  
        records = []
        for arm_name, arm in trial.arms_by_name.items():
            params = arm.parameters
            # What can't I do something like this?
            # mean = arm.metadata[objective_name]
            # sem = arm.metadata['sem'] 
            records.append({
                "arm_name": arm_name,
                "metric_name": self.name,
                "mean": (params["x1"] + 2*params["x2"] - 7)**2 + (2*params["x1"] + params["x2"] - 5)**2,
                "sem": 0.0,
                "trial_index": trial.index,
            })
        return Data(df=pd.DataFrame.from_records(records))
```

**Comment by lena-kashtelyan** [comment]
Hi, @mrthekid! Great questions, let me try to answer them in order. 

> 1) the trial you generate is the one that you evaluate 

This is actually not strictly true; Ax models generate not trials, but `GeneratorRun`-s, which are then added to trials. Typically the suggested workflow is to make any changes to trials before deploying them via a `Runner`, by adding or removing generator runs. Are you using a `Runner`? Is it possible to modify trials before running them in your case?

> 2) that you can compute the objective from the arms somehow

We actually don't assume this either; `Runner.run` attaches metadata to a `Trial` (or `BatchTrial` –– are you using those?), and the idea is that that metadata (`Trial.run_metadata`) can then by used by metrics in `fetch_trial_data`. Frequently, `Trial.run_metadata` will be something like `{"trial_evaluation_job_id": 12345}`, using which metrics can retrieve the evaluation job and its outcomes. 

> I could of course attach some metadata to the trial that has information about all the arms, but this seems inelegant

Do you specifically need to attach metadata about all the arms to the trial or could something like what I suggest above, with recording evaluation job ID in trial metadata and then finding the data your metrics need using those job IDs and arm names (since all arms have unique names)?

**Comment by lena-kashtelyan** [comment]
Did not mean to close the issue above.

**Comment by mrthekid** [comment]
Many thanks for the quick + detailed reply.  

To provide some context - I am just now in the process of migrating from the service to developer API so still trying to figure out exactly what to do.  Anyway, in my test setup I am indeed using a runner and and also attaching metadata via the Runner.run method.   For now, it would probably suffice to attach metadata to the trial only - as opposed to each individual arms.  However, I'm trying to imagine how this might evolve down the road and I was just very unsure as to whether this was the recommended procedure.  

I should stress that my main issue is that I have no good way of associating a job id in the metadata with the configs that ultimately get pushed to prod.  I will merely make a suggestion as to what goes to prod, then there are various other layers of human and software intervention that can change the actual configs.  (In particular, I can't do the modification you suggest in the runner.) I have no control over all these other influences, so 
 I need to think of the suggestions generated by Ax and the actual configs shipped to prod as logically independent.  

In any case, I think I can get around this for the time being by simply adding a new arm and attaching new data to the experiment as it comes in.  To attach the metric mean+ sem to the metadata I end up doing something like:

`trial._run_metadata= {"mean": 10,"sem":2}`

to record the observation since there is no good way to associate that observation with a key.  While this appears to be working fine in my test setup I feel like the API could support this in a more elegant way - or perhaps I'm still just missing something.

**Comment by Balandat** [comment]
On a high level it seems to me that in your setting what you want is not a trial, but take an arm returned by a GeneratorRun, give that to whatever decision pipeline ends up launching some configuration, and then add that configuration as a trial with a custom arm. Otherwise the experiment will have a bunch of proposals that were never actually evaluated (and hence shouldn't really be trials in the first place). Does this make sense?

**Comment by mrthekid** [comment]
Ok, I think I have something along those lines working.  I guess in the end all I really need to update the experiment is:

`experiment.attach_data(observed_data)`

where `observed_data` is a Data object generated however I decide.  

Will go ahead and close.

**Comment by Balandat** [comment]
So I don't think this will work unless you create the appropriate arms on the experiment. Otherwise there is no way for the model to know the parameters for the results of a data object (as the data object does not contain the arm parameterizations). My suggestion would be to create a trial with a custom `GeneratorRun` that contains the arm with the parametrization that ended up being tested, and then also add the data with the results for that arm. 

@lena-kashtelyan does this make sense?

**Comment by mrthekid** [comment]
Yes indeed, I meant that this is what I'm doing now + it is working in my test setup.  I was just a little confused about the best way to get the observed data into the Metric object - though my comment above just meant that I don't really need to do that at all - what matters is the final data.  So, in the example code below I use the metric object for convenience  (and the inelegant `trial._run_metadata = ...` method) though I could just jump straight from generating observed_trial to attaching the data to the experiment if I need to generalize later. 

```

# We imagine that these are just raw observations but for the 
# sake of convenience, here we generate them from the previous suggestions.  
# However, in prod we would like to break the assumption that these
# are from suggestions and get this from various other data sources. 
observed_parameters  = suggestion.arms[0].parameters
observed_pnl = prior_function(observed_parameters)

# Generate a new trial from the observed parameters. 
observed_trial = experiment.new_trial().add_arm(Arm(parameters=observed_parameters))

# We attach the pnl information to the arm as metadata so that we can pick this up and evaluate it
# using the same metric class that we used for the prior. In principle though all we need 
# is a way to generate "new_data" below.
observed_trial._run_metadata = {'pnl': observed_pnl}
observed_trial.mark_running(no_runner_required=True)

# Now, convert the observed trial to the Data format using the metric instance created earlier.
new_data = metric.fetch_trial_data(observed_trial)

# Now, update experiment with the new data.  This is analogous to the attach_trial + completel_trial
# methods in the service API
experiment.attach_data(new_data)
observed_trial.mark_completed()
```

**Comment by lena-kashtelyan** [comment]
> My suggestion would be to create a trial with a custom `GeneratorRun` that contains the arm with the parametrization that ended up being tested, and then also add the data with the results for that arm

@Balandat, yes totally, and the way to create such a trial (with a `GeneratorRun` marked as `MANUAL`) is via `trial.add_arm(...)`.

> I believe I could just jump straight from generating `observed_trial` to attaching the data to the experiment if I need to generalize later

@mrthekid, that sounds like the right move in your case; if there is no way to create "data-fetching logic to obtain data given metadata about the trial run", then there probably is no point in implementing a custom `Metric` and just attaching the data as is done in Service API is the right solution. Something to note about this approach is that you most likely will not be able to use a `GenerationStrategy` to generate candidates in your setup (I can elaborate on why if you need me to), so you'll have to use something like `Models.SOBOL` and then `Models.GPEI` manually to generate candidates for your experiment. Another issue I see with it is that using `Runner.run` for a trial with given arms and then ending up with a potentially different arm deployed is a strange/incorrect setup, where Ax ends up tracking a version of experiment with incorrect trials.

-----

Taking a step back, based on what I'm reading about your setup, it does seem like the correct way of using Ax for it would be this (steps 2 and 3 below are flipped in your current setup, which is what makes Ax tricky to use in your case):
1. Generate a proposed set of arms (via `Models.MY_MODEL(experiment, data).gen` or `generation_strategy.gen`), in the form of a `GeneratorRun` (without adding it as a `Trial` or `BatchTrial` to Ax).
2. Perform the "various other layers of human and software intervention that can change the actual config".
3. Add the arms being evaluated as a result of step 2, to a trial in Ax. This might mean adding the proposed GR from step 1 (in which case you can do `experiment.new_trial(generator_run)` or adding manual arms (created in step 2) via `experiment.new_trial().add_arm(...)`.
4. Run the trial via `Runner.run_trial` (called by `trial.run()`) and attach job run ID to `trial._run_metadata` (assuming now you can get the job run ID, since all modifications to the trial happened in step 2 and you are running the modified trial and recording it in Ax).
5. Go back to step 1, passing in data obtained via `experiment.fetch_data` to `gen` method, since now you should be able to implement your custom `Metric` to fetch the data using the job ID.

Is it possible for you to set up your experiment as to follow these steps? I would highly recommend it, as the way you are currently using Ax is not fully correct and therefore it's hard for us to fully support it.

**Comment by sgbaird** [comment]
#743
